pkg/tenant/topics.go

```diff
@@ -257,30 +262,137 @@ func (s *Server) TopicUpdate(c *gin.Context) {
 	c.JSON(http.StatusOK, t.ToAPI())
 }
 
-// TopicDelete deletes a topic from a user's request with a given ID
-// and returns a 200 OK response instead of an error response.
+// TopicDelete completely destroys a topic, removing the metadata in Trtl and as well
+// as all of the data in Ensign. Because this is irreversible, the first call returns
+// a confirmation token to the user. The user must provide this token in a subsequent
+// request in order to confirm the deletion. Because this operation is asynchronous,
+// the endpoint returns a 202 Accepted response.
 //
 // Route: /topic/:topicID
 func (s *Server) TopicDelete(c *gin.Context) {
 	var (
-		err error
+		err    error
+		ctx    context.Context
+		claims *tokens.Claims
 	)
 
+	// User credentials are required for Quarterdeck requests
+	if ctx, err = middleware.ContextFromRequest(c); err != nil {
+		log.Error().Err(err).Msg("could not create user context from request")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not fetch credentials for authenticated user"))
+		return
+	}
+
+	// User claims are required to verify that the user owns the topic
+	if claims, err = middleware.GetClaims(c); err != nil {
+		log.Error().Err(err).Msg("could not fetch claims from context")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not fetch claims from context"))
+		return
+	}
+
 	// Get the topic ID from the URL and return a 400 response
-	// if the topic does not exist.
+	// if the ID is not parseable
 	var topicID ulid.ULID
 	if topicID, err = ulid.Parse(c.Param("topicID")); err != nil {
-		log.Error().Err(err).Msg("could not parse topic ulid")
-		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse topic ulid"))
+		log.Warn().Err(err).Msg("could not parse topic id")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse topic id"))
+		return
+	}
+
+	// Parse the request body for the confirmation token
+	confirm := &api.Confirmation{}
+	if err = c.BindJSON(confirm); err != nil {
+		log.Warn().Err(err).Msg("could not bind topic delete request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind user request"))
+		return
+	}
+
+	// Sanity check that the ID in the request body matches the ID in the URL
+	if confirm.ID != topicID.String() {
+		log.Warn().Err(err).Msg("topic id in request body does not match topic id in URL")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("id in request body does not match id in URL"))
+		return
+	}
+
+	// Fetch the topic metadata from the database
+	var topic *db.Topic
+	if topic, err = db.RetrieveTopic(ctx, topicID); err != nil {
+		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
```

Comment:
Could we add a check for the error type here; e.g. if the error is `IsNotFound` - I'd prefer not to have a `log.Error()` trigger an alert in Slack but if it's a database error then we should be alerted.